### PR TITLE
feat: Support blackhole, prohibit and unreachable route types 

### PR DESCRIPTION
**Pull request by liangwen12year**
When users have BGP routing setups, it is common practice to blackhole some less-specific routes in order to avoid routing loops, and the BGP router might insert a more specific route dynamically afterwards.

Enhancement:

Reason:

Result:

Issue Tracker Tickets (Jira or BZ if any):

### CODE REVIEW

**Comment by liangwen12year** [comment]
[citest]

**Comment by liangwen12year** [comment]
[citest]

**Comment by liangwen12year** [comment]
[citest]

**Comment by liangwen12year** [comment]
[citest]

**Comment by liangwen12year** [comment]
[citest]

**Comment by liangwen12year** [comment]
[citest bad]

File: tests/playbooks/tests_route_type.yml [hunk 1]
```diff
@@ -0,0 +1,124 @@
+# SPDX-License-Identifier: BSD-3-Clause
+---
+- name: Play for testing route types
+  hosts: all
+  vars:
+    type: veth
+    interface: ethtest0
+
```
Comment (richm) [inline tests/playbooks/tests_route_type.yml:8]: extra line here

Comment (liangwen12year) [reply tests/playbooks/tests_route_type.yml]: removed

**Comment by liangwen12year** [comment]
[citest]

File: README.md [hunk 2]
```diff
@@ -586,13 +586,14 @@ The IP configuration supports the following options:
...
-  `/etc/iproute2/rt_tables.d/*.conf`.
-  Note that Classless inter-domain routing (CIDR) notation or network mask notation
-  are not supported yet.
+  the following entries: `network`, `prefix`, `gateway`, `metric`, `table` and
+  `route_type`. `network` and `prefix` specify the destination network. `table`
+  supports both the numeric table and named table. In order to specify the named
+  table, the users have to ensure the named table is properly defined in
+  `/etc/iproute2/rt_tables` or `/etc/iproute2/rt_tables.d/*.conf`. And the supported
+  `route_type` are `blackhole`, `prohibit`, `unreachable` and these `route_type` routes
+  can not have next hop (gateway) specified. Note that Classless inter-domain routing
```
Comment (tyll) [inline README.md:595]: Please re-rephrase this to not start a new sentence with "and". The documentation needs to clarify that it is also possible not to specify a route_type. Also, it seems that it should be called just "type" instead of "route_type", since it is already clear that the properties are about routes. It would also be nice to link to the type definitions at ip-route here - or some other good source.

File: tests/playbooks/tests_route_type.yml [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                - network: 2001:db8::4
+                  prefix: 128
+                  metric: 2
+                  route_type: blackhole
+                  table: 30600
+                - network: 2001:db8::6
+                  prefix: 128
+                  metric: 4
+                  route_type: prohibit
+                  table: 30600
```
Comment (tyll) [inline tests/playbooks/tests_route_type.yml:68]: It would be great to test also removing some routes.

**Comment by liangwen12year** [comment]
[citest]

File: README.md (hunk 2 above)
Comment (liangwen12year) [reply README.md]: I re-phrased the text and attached the link for the iproute man page.

**Comment by liangwen12year** [comment]
[citest]

**Review by richm — DISMISSED** [review]

**Comment by liangwen12year** [comment]
[citest]

**Comment by liangwen12year** [comment]
[citest]

File: README.md [hunk 4]
```diff
@@ -586,13 +586,16 @@ The IP configuration supports the following options:
...
-  Note that Classless inter-domain routing (CIDR) notation or network mask notation
-  are not supported yet.
+  numeric table and named table. In order to specify the named table, the users have to
+  ensure the named table is properly defined in `/etc/iproute2/rt_tables` or
+  `/etc/iproute2/rt_tables.d/*.conf`. The supported `type` are `blackhole`, `prohibit`,
+  `unreachable` and these three type routes can not have next hop (gateway) specified.
+  It is possible to create a route without `type` specified. For the route `type`
+  definition, please refer to [man 8 ip-route](https://man7.org/linux/man-pages/man8/ip-route.8.html#DESCRIPTION).
+  Note that Classless inter-domain routing(CIDR) notation or network mask notation are
+  not supported yet.
```
Comment (tyll) [inline README.md:598]: The text does not flow. Here is another proposal:

```suggestion
  `/etc/iproute2/rt_tables.d/*.conf`. The optional `type` key supports the values `blackhole`, `prohibit`,
  and `unreachable`, see [man 8 ip-route](https://man7.org/linux/man-pages/man8/ip-route.8.html#DESCRIPTION) for their definition. Routes with these types do not support a gateway. If the type is not specified, the route is considered a unicast route.
  
 Note that the classless inter-domain routing(CIDR) notation or the network mask notation are
  not supported for the `network` key.
```

Is the statement regarding unicast routes correct?
Please

File: module_utils/network_lsr/argument_validator.py [hunk 5]
```diff
@@ -688,13 +693,20 @@ def _validate_post(self, value, name, result):
...
             if family != gateway["family"]:
                 raise ValidationError(
                     name,
                     "conflicting address family between network and gateway '%s'"
                     % (gateway["address"]),
                 )
+            if type is not None:
+                raise ValidationError(
+                    name,
+                    "a %s route can not have a next hop '%s'"
```
Comment (tyll) [inline module_utils/network_lsr/argument_validator.py:707]: This is called gateway in the role, not next hop.

File: module_utils/network_lsr/argument_validator.py [hunk 6]
```diff
@@ -688,13 +693,20 @@ def _validate_post(self, value, name, result):
         result["family"] = family
 
         gateway = result["gateway"]
+        type = result["type"]
```
Comment (tyll) [inline module_utils/network_lsr/argument_validator.py:696]: `type` is a builtin function, so it should not be assigned: https://docs.python.org/3/library/functions.html#type

It is strange that this is not reported by a python linter or code checker, for example:
https://pylint.pycqa.org/en/latest/user_guide/messages/warning/redefined-builtin.html

File: tests/ensure_provider_tests.py [hunk 7]
```diff
@@ -95,6 +95,11 @@
     },
     "playbooks/tests_reapply.yml": {},
     "playbooks/tests_route_table.yml": {},
+    "playbooks/tests_route_type.yml": {
+        MINIMUM_VERSION: "'1.36.0'",
```
Comment (tyll) [inline tests/ensure_provider_tests.py:99]: What happens when the test runs against an older version of NM? It does not seem that the module protects against this.

File: tests/playbooks/tests_route_type.yml [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+
+    - name: Remove routes in table 30200
+      import_role:
+        name: linux-system-roles.network
+      vars:
+        network_state:
+          routes:
+            config:
+              - table-id: 30200
+                state: absent
```
Comment (tyll) [inline tests/playbooks/tests_route_type.yml:112]: Please remove them using `network_connections` since this is the code path that you changed.

File: tests/ensure_provider_tests.py (hunk 7 above)
Comment (liangwen12year) [reply tests/ensure_provider_tests.py:99]: right, I should add an validation function.

File: README.md (hunk 4 above)
Comment (liangwen12year) [reply README.md:598]: I think that the statement regarding the unicast route is correct.

Comment (tyll) [reply README.md:598]: I do not see a reaction to my proposal, therefore I unresolved this conversation.

File: module_utils/network_lsr/argument_validator.py (hunk 6 above)
Comment (tyll) [reply module_utils/network_lsr/argument_validator.py:696]: This is not resolved.

File: module_utils/network_lsr/argument_validator.py (hunk 5 above)
Comment (tyll) [reply module_utils/network_lsr/argument_validator.py:707]: This is not resolved.

File: README.md (hunk 4 above)
Comment (liangwen12year) [reply README.md]: Sorry, I forgot to push the changes last night.

**Comment by liangwen12year** [comment]
[citest]

File: README.md (hunk 4 above)
Comment (liangwen12year) [reply README.md]: Your proposal looks quite nice to me, Thanks.

**Comment by liangwen12year** [comment]
@tyll , Thanks for the comments, I think that I addressed them all.

File: module_utils/network_lsr/nm/provider.py [hunk 9]
```diff
@@ -65,6 +65,18 @@ def get_client_version(self):
         nm_client = client.get_client()
         return nm_client.get_version()
 
+    def nm_support_version_cmp(self, nm_support_version):
+        v = tuple(self.get_client_version().split("."))
```
Comment (tyll) [inline module_utils/network_lsr/nm/provider.py:69]: Please do not introduce new single letter variables.

Comment (liangwen12year) [reply module_utils/network_lsr/nm/provider.py]: I replaced with meaningful variable, thanks.

**Comment by liangwen12year** [comment]
[citest]

File: README.md [hunk 10]
```diff
@@ -586,13 +586,16 @@ The IP configuration supports the following options:
 
   Static route configuration can be specified via a list of routes given in the
```
Comment (tyll) [inline README.md:587]: regarding the commit message, seems the title can be simplified as:

"feat: Support blackhole, prohibit and unreachable route types"

I am pretty sure the "types" needs to be plural.

Regarding the details in the commit message, it is missing that this is only for the nm provider using network_connections. Is it already supported via network_state, then it could mention this as well. The detailed explanations for the different types are hard to understand. I am not convinced that it needs to be in that detail as well. If you want that detail, please make it easier to understand. Otherwise the key message is that the types are useful in BGP setups which could be enough.

If we have a Jira ticket for this, maybe it is also good practice to mention it as resolves.

File: module_utils/network_lsr/nm/provider.py [hunk 11]
```diff
@@ -65,6 +65,18 @@ def get_client_version(self):
...
         return nm_client.get_version()
 
+    def nm_support_version_cmp(self, nm_support_version):
+        client_ver = tuple(self.get_client_version().split("."))
+        major, minor, micro = map(int, client_ver)
+        nm_support_ver = tuple(nm_support_version.split("."))
+        nm_major, nm_minor, nm_micro = map(int, nm_support_ver)
+        # emulate the behavior of `nm_client_get_version_info()` to encode NM daemon
+        # version as "(major << 16 | minor << 8 | micro)", but
+        # `nm_client_get_version_info()` is only supported to use since NM 1.42
```
[the remark this answers is not in the document]
Comment (liangwen12year) [reply module_utils/network_lsr/nm/provider.py:75]: I think that it is a good idea.

**Comment by liangwen12year** [comment]
[citest]

**Comment by liangwen12year** [comment]
[citest]

File: library/network_connections.py [hunk 12]
```diff
@@ -2193,6 +2197,20 @@ def run_action_up(self, idx):
 ###############################################################################
 
 
+def get_client_version_tuple(client_version):
```
Comment (tyll) [inline library/network_connections.py:2200]: Suggested new name:

`version_to_tuple(version)`

This is not specific for the client version but works for any dot-separated version string.

File: README.md [hunk 13]
```diff
@@ -586,13 +586,16 @@ The IP configuration supports the following options:
...
   `network` and `prefix` specify the destination network. `table` supports both the
-  numeric table and named table. In order to specify the named table, the users have
-  to ensure the named table is properly defined in `/etc/iproute2/rt_tables` or
-  `/etc/iproute2/rt_tables.d/*.conf`.
-  Note that Classless inter-domain routing (CIDR) notation or network mask notation
-  are not supported yet.
+  numeric table and named table. In order to specify the named table, the users have to
+  ensure the named table is properly defined in `/etc/iproute2/rt_tables` or
+  `/etc/iproute2/rt_tables.d/*.conf`. The optional `type` key supports the values
+  `blackhole`, `prohibit`, and `unreachable`, see [man 8 ip-route](https://man7.org/linux/man-pages/man8/ip-route.8.html#DESCRIPTION)
```
Comment (tyll) [inline README.md:594]:
```suggestion
  `blackhole`, `prohibit`, and `unreachable`.
  See [man 8 ip-route](https://man7.org/linux/man-pages/man8/ip-route.8.html#DESCRIPTION)
```

**Comment by tyll** [comment]
Only minor remarks from me, I don't need to re-review this. I am fine with someone else approving this after the changes.

**Comment by liangwen12year** [comment]
[citest]

**Review by richm — APPROVED** [review]
